### PR TITLE
Allow including both podSelector and namespaceSelector in a NetworkPolicyPeer

### LINKED ISSUE #58637: Proposal: NetworkPolicy support for selecting pods in another namespace
People want the ability to specify specific pods within a (different) namespace in NetworkPolicies.

I propose we change the documentation/definition of `NetworkPolicyPeer` so that instead of saying "Exactly one of its fields must be specified", we say that there are 4 valid kinds of NetworkPolicyPeer:

1. If only `PodSelector` is set, then the peer selects the indicated pods in the same namespace as the NetworkPolicy.
2. If only `NamespaceSelector` is set, then the peer selects all pods in the indicated namespaces.
3. (**NEW**) If both `PodSelector` and `NamespaceSelector` are set, then the peer selects the indicated pods in the indicated namespaces. (And, in particular, if `PodSelector` is set and `NamespaceSelector` is `{}`, then the peer selects the indicated pods in all namespaces.)
4. If only `IPBlock` is set, then the peer selects the indicated IP addresses.

As before, it would be invalid to have a `NetworkPolicyPeer` in which all fields were unset, or in which both `IPBlock` and any other field were set.

This shouldn't create any compatibility issues: no old policies would be interpreted differently under the new rules, because the new rules only describe the interpretation of policies that would be invalid under the old rules. And likewise, policies using the new features would not be interpreted incorrectly by old plugins; they'd just be rejected as invalid.

/sig network
/kind feature

### PR DESCRIPTION
**Pull request by danwinship**
Updates NetworkPolicy API docs and validation to allow podSelector and namespaceSelector to be specified together in a NetworkPolicyPeer

Fixes #58637

**Release note**:
```release-note
NetworkPolicies can now target specific pods in other namespaces by including both a namespaceSelector and a podSelector in the same peer element.
```

### CODE REVIEW

**Comment by danwinship** [comment]
/sig network

Is there anything else I missed in kubernetes itself? We'll also need a docs PR and a kubectl PR...

**Comment by danwinship** [comment]
Having actually written the docs now, I'm maybe more convinceable that we should go with `podAndNamespaceSelector` instead; it would work more cleanly with the idea of separately documenting each field in `NetworkPolicyPeer` like we do...

**Comment by caseydavenport** [comment]
> Having actually written the docs now, I'm maybe more convinceable that we should go with podAndNamespaceSelector instead; it would work more cleanly with the idea of separately documenting each field in NetworkPolicyPeer like we do...

I'd prefer to not introduce a new field such as `podAndNamespaceSelector`

As a user of the API, I think the AND of `podSelector` and `namespaceSelector` is exactly how I'd expect those two to behave when both specified. In fact, even a generic AND would make sense to me, though I don't see what use-case I'd have for it.

As an implementor, it's almost no work on my side to simply support the AND combination of pod / namespace, whereas there's more for me to do to support a new selector field.

**Comment by thockin** [comment]
My concern is that the AND isn't generic (unless we make it so) and is
therefore vastly more confusing.

**Comment by caseydavenport** [comment]
> My concern is that the AND isn't generic (unless we make it so) and is
therefore vastly more confusing.

That may be, though with a well written out validation error I'd guess it would be fairly clear.

I understand it might be extra work for other implementors to do the generic AND, but from a Calico perspective it's trivial. So personally I'd rather have us implement the generic AND than add an extra type for people to reason with.

**Comment by danwinship** [comment]
> That may be, though with a well written out validation error I'd guess it would be fairly clear.

"may not specify both ipBlock and another peer"

**Comment by thockin** [comment]
I am OK either way.  Let's just get it in.

**Comment by Lion-Wei** [comment]
I like this, is this anything I can help? Maybe the kubectl pr? @danwinship

**Comment by danwinship** [comment]
@Lion-Wei: sure. "kubectl describe" needs to be extended to be able to describe policies where a NetworkPolicyPeer has both podSelector and namespaceSelector set (describing it according to the semantics seen in the API docs here).

File: hack/make-rules/update.sh [hunk 1]
```diff
@@ -47,7 +47,7 @@ if ! ${ALL} ; then
 	echo "Running in short-circuit mode; run with FORCE_ALL=true to force all scripts to run."
 fi
 
-"${KUBE_ROOT}/hack/godep-restore.sh" ${V}
+#"${KUBE_ROOT}/hack/godep-restore.sh" ${V}
```
Comment (caseydavenport) [inline hack/make-rules/update.sh]: Need to add this back?

**Review by caseydavenport — COMMENTED** [review]
This LGTM - one minor comment.

I'm in favor of merging as-is, and further relaxing the validation if we get feedback that it's confusing.

**Comment by Lion-Wei** [comment]
@danwinship I will push that pr as soon as this got merged. : )

**Comment by danwinship** [comment]
@caseydavenport repushed with the stray update.sh change removed if you want to /lgtm it

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60452)

**Comment by oivindoh** [comment]
What's holding this thing of **outstanding natural beauty** back?

**Comment by danwinship** [comment]
It's waiting for 1.10 to go out and master to reopen for 1.11. I'm not sure where the /lgtm went.

**Comment by ahmetb** [comment]
Is this technically a breaking change in the behavior? (or was it impossible to specify both namespaceSelector and podSelector on the same peer spec?)

**Comment by oivindoh** [comment]
@caseydavenport does this require a change in calico to work?

**Comment by danwinship** [comment]
@ahmetb It was previously impossible to specify both; validation would reject the resource in that case

**Pull request merged**